### PR TITLE
temporary fix for heredocs

### PR DESCRIPTION
**Pull request by UziTech**
### Description of the Change

Temporary fix for heredocs explained in [#77 (comment)](https://github.com/atom/language-php/pull/77#issuecomment-140815381)

### Alternate Designs

#77 has been waiting on a fix for over a year on another package that doesn't look like it is being worked on. Fixing the underlying problem would be the optimal solution but this will provide a fix in the mean time.

### Benefits

Heredocs will work as expected including sub-styling with `<<<HTML`, `<<<SQL`, etc.

### Possible Drawbacks

None that I can see.

### Applicable Issues

#10, #57, #63, #78, #113, #133, #155

### CODE REVIEW

**Comment by dustinwilson** [comment]
Thanks for doing this pull request. I tested this, and interpolation isn't working.

Because of the changes you made the interpolation stuff at [line 475](https://github.com/UziTech/language-php/blob/cc576387c51e89197e964e4aabfb89be4eeb6a32/grammars/php.cson#L475) doesn't do anything.

However, if you put the interpolation stuff like so after [line 713](https://github.com/UziTech/language-php/blob/cc576387c51e89197e964e4aabfb89be4eeb6a32/grammars/php.cson#L713) it should work (and probably for the other ones if you repeat it):

```CoffeeScript
{
  'begin': '(<<<)\\s*([\'"]?)([a-zA-Z_]+[a-zA-Z0-9_]*)(\\2)'
  'beginCaptures':
    '1':
      'name': 'punctuation.definition.string.php'
    '3':
      'name': 'keyword.operator.heredoc.php'
  'end': '^(\\3)\\b'
  'endCaptures':
    '1':
      'name': 'keyword.operator.heredoc.php'
  'patterns': [
    {
      'include': '#interpolation'
    }
  ]
}
```

**Comment by UziTech** [comment]
Thank you @dustinwilson. I noticed that last night but I wasn't sure how to fix it.

**Comment by UziTech** [comment]
The only other part I'm not 100% sure on is in the HTML heredoc test it looks like `<strong>rainbows</strong>` just gets the `text.html` scope as if it is a string.

In the other pull request (#77) it looks like the tag should have additional scopes

I'm not sure what to do (if anything) about that.

Everything seems to actually work in Atom

**Comment by UziTech** [comment]
It looks like the reason it is not adding additional scopes to the tag is because the 'language-html' package is not activated in the tests.

That should be fine as long as it gets the `text.html` scope the 'language-html' package should take over in atom so there should be no need to test that in this package

**Comment by 50Wliu** [comment]
In that test, add a waitsForPromise like follows:
```coffee
waitsForPromise ->
  atom.packages.activatePackage('language-html')
```

The rest of the test will need to be wrapped in a `runs` function, but then it should work.

**Comment by UziTech** [comment]
@50Wliu Do you think we need to add the language-html scopes to the test?

I could see an issue where language-html changes something that breaks these tests and it would be hard to figure out why they broke.

I do test for the `text.html` scope on anything that is supposed to be html. Adding the `text.html` scope should be enough for this package.

**Comment by 50Wliu** [comment]
My worry is that what if someone removes the HTML include later on?  Then we have no way of knowing that it broke just from the specs.

I wholeheartedly agree that the whole cross-language-specs thing is very fragile though - if you have a potential solution, I would love to hear it!

**Comment by UziTech** [comment]
True. I don't think we are going to get around this one.

Do you want me to add tests for the other languages as well?

**Comment by 50Wliu** [comment]
Yes please.  Just a simple test will suffice.

**Comment by UziTech** [comment]
lol apparently this is happening for language-css.

The tests for css can't pass because the stable channel has a different scope for curly braces than the beta channel.

**Comment by 50Wliu** [comment]
Create the tests so that they pass using the latest version of the language package (looks like that's what you did).

**Comment by UziTech** [comment]
you want them to pass the latest stable version? Then they will continue to be broken.

If I change them to pass the beta version then they will at least pass when the beta version becomes stable

**Comment by 50Wliu** [comment]
Yes, what you did is fine.  When I say latest, I mean "the latest version available", not "the version available on stable".

File: grammars/php.cson [hunk 1]
```diff
@@ -470,15 +470,8 @@
   'heredoc':
     'patterns': [
       {
-        'begin': '(?=<<<\\s*("?)([a-zA-Z_]+[a-zA-Z0-9_]*)(\\1)\\s*$)'
+        'begin': '(?=<<<\\s*("?)([a-zA-Z_]+[a-zA-Z0-9_]*)(\\1)[ \\t]*$)'
```
Comment (50Wliu) [inline grammars/php.cson]: Why was this changed?

File: grammars/php.cson [hunk 2]
```diff
@@ -487,233 +480,506 @@
...
           }
         ]
       }
     ]
-    'repository':
-      'heredoc_interior':
+  'heredoc_interior':
+    'patterns': [
+      {
+        'begin': '(<<<)\\s*("?)(HTML)(\\2)([ \\t]+)?$'
```
Comment (50Wliu) [inline grammars/php.cson]: The `[ \\t]` should be changed to use `\\s` instead.

File: spec/php-spec.coffee [hunk 3]
```diff
@@ -488,6 +488,690 @@ describe 'PHP grammar', ->
...
+        expect(tokens[9][0]).toEqual value: '    console.log(x*x);', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.nowdoc.php']
+        expect(tokens[10][0]).toEqual value: '  });', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.nowdoc.php']
+        expect(tokens[11][0]).toEqual value: '}', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.nowdoc.php']
+        expect(tokens[12][0]).toEqual value: 'GITHUB', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'string.unquoted.heredoc.nowdoc.php', 'keyword.operator.heredoc.php']
+        expect(tokens[12][1]).toEqual value: ';', scopes: ['text.html.php', 'meta.embedded.block.php', 'source.php', 'punctuation.terminator.expression.php']
+
+      it 'should tokenize a heredoc with embedded HTML and interpolation correctly', ->
+        waitsForPromise ->
+          atom.packages.activatePackage('language-html')
+        runs ->
```
Comment (50Wliu) [inline spec/php-spec.coffee]: Add a newline before the runs

File: grammars/php.cson [hunk 4]
```diff
@@ -487,233 +480,506 @@
         ]
       }
       {
-        'begin': '(?=<<<\\s*(\'?)([a-zA-Z_]+[a-zA-Z0-9_]*)(\\1)\\s*$)'
+        'begin': '(?=<<<\\s*\'([a-zA-Z_]+[a-zA-Z0-9_]*)\'[ \\t]*$)'
         'end': '(?!\\G)'
         'name': 'string.unquoted.heredoc.nowdoc.php'
```
Comment (50Wliu) [inline grammars/php.cson]: Shouldn't this just be `string.unquoted.nowdoc.php` instead?

File: grammars/php.cson (hunk 1 above)
Comment (UziTech) [inline grammars/php.cson]: according to the [original pull request](https://github.com/atom/language-php/pull/77/commits/eb4ae2656bd8a427d2026f082ba2c62da3d62654#diff-9972c3dda8fb0e2f637ca073c25e7e6dR439)
```
[ \\t] was used instead of \\s because \\s will match a newline
```

File: grammars/php.cson (hunk 4 above)
Comment (UziTech) [inline grammars/php.cson]: that's what I was thinking but I didn't know if there was a reason for the `.heredoc` (styling or whatever)

I can change it if you want

File: grammars/php.cson (hunk 2 above)
Comment (UziTech) [inline grammars/php.cson]: see https://github.com/atom/language-php/pull/184#discussion_r105465668

**Comment by UziTech** [comment]
Seriously this is a never ending battle. 

Now `language-json` scopes changed in beta. 

I'm thinking testing the other packages' scopes is not going to be worth the hassle of having to update the tests any time one of the other language packages updates.

File: grammars/php.cson (hunk 1 above)
Comment (50Wliu) [inline grammars/php.cson]: I highly doubt that as the begin pattern can only match a single line.  If you change it, do the specs still pass?

Comment (UziTech) [inline grammars/php.cson]: ya everything looks fine when it is `\\s`

**Comment by 50Wliu** [comment]
> I'm thinking testing the other packages' scopes is not going to be worth the hassle of having to update the tests any time one of the other language packages updates.

Maybe we could test just one of the scopes (probably meta ones) that isn't likely to change using `.toContain`.

**Comment by UziTech** [comment]
I changed the tests to check for the token value and any scopes that were added by language-php. That way if the included languages are removed then the token values won't match but it won't fail if other packages change their scopes only if they tokenize the string differently, which is unlikely.

File: grammars/php.cson [hunk 5]
```diff
@@ -487,233 +480,506 @@
...
+            'name': 'keyword.operator.heredoc.php'
+          '3':
+            'name': 'invalid.illegal.trailing-whitespace.php'
+        'contentName': 'text.html'
+        'end': '^(\\2)\\b'
+        'endCaptures':
+          '0':
+            'name': 'punctuation.section.embedded.end.php'
+          '1':
+            'name': 'keyword.operator.heredoc.php'
```
Comment (UziTech) [inline grammars/php.cson]: do you want me to change this to `keyword.operator.nowdoc.php`

Comment (50Wliu) [inline grammars/php.cson]: (assuming you meant for `nowdoc_interior` and not `heredoc_interior`) yes please.

File: grammars/php.cson [hunk 6]
```diff
@@ -487,233 +480,506 @@
...
           }
         ]
       }
     ]
-    'repository':
-      'heredoc_interior':
+  'heredoc_interior':
+    'patterns': [
+      {
+        'begin': '(<<<)\\s*("?)(HTML)(\\2)(\\s+)?$'
```
Comment (50Wliu) [inline grammars/php.cson]: One last thing.  I believe that the `(\\s+)?` can be changed to simply `(\\s*)` - can you confirm?

**Comment by 50Wliu** [comment]
Thanks!

**Comment by 50Wliu** [comment]
Current ETA is Atom 1.17.0.  You can manually install language-php using `apm install language-php` though that is **highly** discouraged (and will display deprecation warnings starting with Atom 1.16.0).

**Comment by timbotron** [comment]
@50Wliu Thank you for the info! I think I'll just wait. :)

**Comment by AndreasWebdev** [comment]
It seems like this was not fixed. I am on 1.17.2

image

**Comment by jens1o** [comment]
The closing character must not be indented, so make sure `FDF;` is at the very first column.
